transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+        {
+           control->iocbGet = 0;
+        }
+    }
+    else
+    {
+        #ifdef DEBUG
+            fprintf (stderr, "Could not find iocb\n");
+        #endif
+    }
+
+    pthread_mutex_unlock(&(control->iocbLock));
+    return iocb;
+}
+
+/**
+  Put an iocb back on the pool of IOCBs
+*/
+static inline void putIOCB(struct io_control * control, struct iocb * iocbBack) {
+    pthread_mutex_lock(&(control->iocbLock));
+
+    #ifdef DEBUG
+       fprintf (stderr, "putIOCB::used=%d, queueSize=%d, get=%d, put=%d\n", control->used, control->queueSize, control->iocbGet, control->iocbPut);
+    #endif
+
+    control->used--;
+    control->iocb[control->iocbPut++] = iocbBack;
+    if (control->iocbPut >= control->queueSize)
+    {
+       control->iocbPut = 0;
+    }
+    pthread_mutex_unlock(&(control->iocbLock));
+}
+
+static inline void * getBuffer(JNIEnv* env, jobject pointer ) {
+    return (*env)->GetDirectBufferAddress(env, pointer);
+}
+
+JNIEXPORT jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newContext(JNIEnv* env, jclass clazz, jint queueSize) {
+
+    io_context_t libaioContext;
+    int i = 0;
+
+    int res = io_queue_init(queueSize, &libaioContext);
+    if (res) {
+        // Error, so need to release whatever was done before
+        free(libaioContext);
+
+        throwRuntimeException(env, exceptionMessage("Cannot initialize queue:", -res));
+        return NULL;
+    }
+
+    struct iocb ** iocb = (struct iocb **)malloc(sizeof(struct iocb *) * queueSize);
+
+    for (i = 0; i < queueSize; i++)
+    {
+       iocb[i] = (struct iocb *)malloc(sizeof(struct iocb));
+    }
+
+    struct io_event * events = (struct io_event *)malloc(sizeof(struct io_event) * queueSize);
+
+    struct io_control * theControl = (struct io_control *) malloc(sizeof(struct io_control));
+
+    theControl->errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (theControl->errorInfoClass == NULL)
+    {
+       return;
+    }
+
+
+    /// The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    theControl->errorInfoClass = (jclass)(*env)->NewGlobalRef(env, (jobject)(theControl->errorInfoClass));
+
+    theControl->errorInfoConstr = (*env)->GetMethodID(env, theControl->errorInfoClass, "<init>", "(Ljava/lang/Object;ILjava/lang/String;)V");
+
+    if (theControl->errorInfoConstr == NULL)
+    {
+       return;
+    }
+
+    /// The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    theControl->errorInfoConstr = (jmethodID)(*env)->NewGlobalRef(env, (jobject)(theControl->errorInfoConstr));
+
+    theControl->ioContext = libaioContext;
+    theControl->events = events;
+    theControl->iocb = iocb;
+    theControl->queueSize = queueSize;
+    theControl->iocbPut = 0;
+    theControl->iocbGet = 0;
+    theControl->used = 0;
+    pthread_mutex_init(&(theControl->iocbLock), 0);
```

Comment:
I see,

I will move the pthread_mutex_init at the begining and delete the structure if failed. (as well as throw an error).